### PR TITLE
SAK-31130 - improved performance for retrieving submissions on the Grade Report tab of Assignments

### PR DESCRIPTION
**Pull request by bbailla2**
Improve the loading time

Note when testing:
If you create a new assignment, visit the 'Grades' page before visiting the 'Grade Report' tab - the first time you view grades, dummy submission objects are created for all the assignments, so it will be slow no matter what. This ticket is only concerned with speeding up the retrieval of the submission objects on the Grade Report tab

### CODE REVIEW

**Comment by ottenhoff** [comment]
@jonespm thoughts on this one?

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java [hunk 1]
```diff
@@ -3514,6 +3515,48 @@ public AssignmentSubmission getSubmission(String assignmentReference, User perso
...
+		{
+			if (a.isGroup())
+			{
+				try
+				{
+					Site _site = SiteService.getSite(a.getContext());
+					for (User user : users)
+					{
+						AssignmentSubmission submission = null;
+						Collection<Group> groups = (Collection<Group>) _site.getGroupsWithMember(user.getId());
```
Comment (jonespm) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java]: I'm looking at getSubmission(String assignmentReference, User person) and it looks like it already handles this loop. I feel like there's unnecessary duplication in this block in the grouped case.

**Comment by jonespm** [comment]
I think this looks like a good idea, changing some of the queries for ungrouped assignments. I commented on a little code duplication is all, but the rest of it looks okay to me.

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java (hunk 1 above)
Comment (ottenhoff) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java]: @bbailla2 thoughts on this comment?

Comment (bbailla2) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java]: Yeah, they're duplicated. I'll refactor and push soon

File: assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java [hunk 2]
```diff
@@ -3514,6 +3499,76 @@ public AssignmentSubmission getSubmission(String assignmentReference, User perso
...
+		Map<User, AssignmentSubmission> userSubmissionMap = new HashMap<>();
+		if (a != null && !CollectionUtils.isEmpty(users))
+		{
+			if (a.isGroup())
+			{
+				userSubmissionMap.putAll(getUserGroupSubmissionMap(a, users));
+			}
+			else
+			{
+				// SAK-31130 - get all submissions for these users with a single query
```
Comment (bjones86) [inline assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java]: You don't need to reference the SAK ticket number in comments. Git blame can tell us this info. Same deal in a few other places on this PR.

**Comment by bjones86** [comment]
I think this is ready to be merged now

**Pull request merged**